app/src/main/java/org/schabi/newpipe/player/mediabrowser/MediaBrowserConnector.kt

```diff
@@ -0,0 +1,814 @@
+import org.schabi.newpipe.database.playlist.PlaylistLocalItem
+import org.schabi.newpipe.database.playlist.PlaylistStreamEntry
+import org.schabi.newpipe.database.playlist.model.PlaylistRemoteEntity
+import org.schabi.newpipe.error.ErrorInfo
+import org.schabi.newpipe.error.UserAction
+import org.schabi.newpipe.extractor.InfoItem
+import org.schabi.newpipe.extractor.InfoItem.InfoType
+import org.schabi.newpipe.extractor.ListInfo
+import org.schabi.newpipe.extractor.channel.ChannelInfo
+import org.schabi.newpipe.extractor.channel.ChannelInfoItem
+import org.schabi.newpipe.extractor.exceptions.ContentNotAvailableException
+import org.schabi.newpipe.extractor.exceptions.ContentNotSupportedException
+import org.schabi.newpipe.extractor.linkhandler.ListLinkHandler
+import org.schabi.newpipe.extractor.playlist.PlaylistInfo
+import org.schabi.newpipe.extractor.playlist.PlaylistInfoItem
+import org.schabi.newpipe.extractor.search.SearchExtractor.NothingFoundException
+import org.schabi.newpipe.extractor.search.SearchInfo
+import org.schabi.newpipe.extractor.stream.StreamInfo
+import org.schabi.newpipe.extractor.stream.StreamInfoItem
+import org.schabi.newpipe.local.bookmark.MergedPlaylistManager
+import org.schabi.newpipe.local.playlist.LocalPlaylistManager
+import org.schabi.newpipe.local.playlist.RemotePlaylistManager
+import org.schabi.newpipe.player.PlayerService
+import org.schabi.newpipe.player.playqueue.ChannelTabPlayQueue
+import org.schabi.newpipe.player.playqueue.PlayQueue
+import org.schabi.newpipe.player.playqueue.PlaylistPlayQueue
+import org.schabi.newpipe.player.playqueue.SinglePlayQueue
+import org.schabi.newpipe.util.ChannelTabHelper
+import org.schabi.newpipe.util.ExtractorHelper
+import org.schabi.newpipe.util.NavigationHelper
+import org.schabi.newpipe.util.ServiceHelper
+import java.lang.Exception
+import java.lang.IllegalStateException
+import java.lang.NullPointerException
+import java.util.ArrayList
+import java.util.stream.Collectors
+
+class MediaBrowserConnector(private val playerService: PlayerService) : PlaybackPreparer {
+    private val context: Context = playerService
+    private val sessionConnector: MediaSessionConnector
+    private val mediaSession: MediaSessionCompat = MediaSessionCompat(playerService, TAG)
+
+    private val database: AppDatabase
+        get() = NewPipeDatabase.getInstance(context)
+    private val mergedPlaylists
+        get() = MergedPlaylistManager.getMergedOrderedPlaylists(
+            LocalPlaylistManager(database),
+            RemotePlaylistManager(database)
+        )
+    private var prepareOrPlayDisposable: Disposable? = null
+    private var searchDisposable: Disposable? = null
+
+    var bookmarksNotificationsDisposable: Disposable
+
+    init {
+        sessionConnector = MediaSessionConnector(mediaSession)
+        sessionConnector.setMetadataDeduplicationEnabled(true)
+        sessionConnector.setPlaybackPreparer(this)
```

Comment:
I don’t understand how this is intended to work, the `MediaSessionConnector` is instantiated here (as opposed to `MediaSessionPlayerUI` before) and we set some `PlaybackPreparer` interface that is related to the Auto lifecycle.

But then we pass it again to `MediaSessionPlayerUI` (as before) where a lot of things related to the normal media session are added in addition to the `PlaybackPreparer`.

This sounds like we could end in a situation where multiple parts of NewPipe fight for who handles Playback preparation, if not now then maybe later when more features are added relating to the media session.

Reply:
I'm not sure what's the best way to address this. Perhaps other features that require playback preparation could also go through the same media browser interface?